http/server_test.ts

```diff
@@ -298,4 +303,99 @@ test(async function writeStringReaderResponse(): Promise<void> {
+    localAddr: "",
+    remoteAddr: "",
+    rid: -1,
+    closeRead: (): void => {},
+    closeWrite: (): void => {},
+    read: async (): Promise<Deno.ReadResult> => {
+      return { eof: true, nread: 0 };
+    },
+    write: async (): Promise<number> => {
+      return -1;
+    },
+    close: (): void => {}
+  };
+}
+
+test(async function readRequestError(): Promise<void> {
+  let input = `GET / HTTP/1.1
+malformedHeader
+`;
+  const buf = new Buffer(enc.encode(input));
+  const reader = new BufReader(buf);
+  const conn = createConnMock();
+  const [_, err] = await readRequest(conn, reader);
+  const e: any = err; // eslint-disable-line @typescript-eslint/no-explicit-any
+  assert(e instanceof Error);
+  assertEquals(e.message, "Unable to proceed request");
+});
+
+// Port from Go
+// https://github.com/golang/go/blob/master/src/net/http/request_test.go
+test(async function testReadRequestError(): Promise<void> {
+  const testCases = {
+    0: {
+      in: "GET / HTTP/1.1\r\nheader: foo\r\n\r\n",
+      headers: [{ key: "header", value: "foo" }],
+      err: null
+    },
+    1: { in: "GET / HTTP/1.1\r\nheader:foo\r\n", err: "EOF", headers: [] },
+    2: { in: "", err: "EOF", headers: [] },
+    // 3: {
+    //   in: "HEAD / HTTP/1.1\r\nContent-Length:4\r\n\r\n",
+    //   err: "http: method cannot contain a Content-Length"
+    // },
+    4: {
+      in: "HEAD / HTTP/1.1\r\n\r\n",
+      headers: [],
+      err: null
+    }
+    // Multiple Content-Length values should either be
+    // deduplicated if same or reject otherwise
+    // See Issue 16490.
+    // 5: {
+    //   in:
+    //     "POST / HTTP/1.1\r\nContent-Length: 10\r\nContent-Length: 0\r\n\r\nGopher hey\r\n",
+    //   err: "cannot contain multiple Content-Length headers"
+    // },
+    // 6: {
+    //   in:
+    //     "POST / HTTP/1.1\r\nContent-Length: 10\r\nContent-Length: 6\r\n\r\nGopher\r\n",
+    //   err: "cannot contain multiple Content-Length headers"
+    // },
+    // 7: {
+    //   in:
+    //     "PUT / HTTP/1.1\r\nContent-Length: 6 \r\nContent-Length: 6\r\nContent-Length:6\r\n\r\nGopher\r\n",
+    //   err: null,
+    //   headers: [{ key: "Content-Length", value: "6" }]
+    // },
+    // 8: {
+    //   in: "PUT / HTTP/1.1\r\nContent-Length: 1\r\nContent-Length: 6 \r\n\r\n",
+    //   err: "cannot contain multiple Content-Length headers"
+    // },
+    // 9: {
+    //   in: "POST / HTTP/1.1\r\nContent-Length:\r\nContent-Length: 3\r\n\r\n",
+    //   err: "cannot contain multiple Content-Length headers"
+    // },
+    // 10: {
+    //   in: "HEAD / HTTP/1.1\r\nContent-Length:0\r\nContent-Length: 0\r\n\r\n",
+    //   headers: [{ key: "Content-Length", value: "0" }],
+    //   err: null
+    // }
+  };
+  for (const p in testCases) {
+    const test = testCases[p];
+    const buf = new Buffer(enc.encode(test.in).buffer);
+    const reader = new BufReader(buf);
+    const conn = createConnMock();
+    const [req, err] = await readRequest(conn, reader);
```

Comment:
In the Go code, it looks like this:
```
		req, err := ReadRequest(bufio.NewReader(strings.NewReader(tt.in)))
```
I think you can do the exact same thing here:
```ts
import { StringReader } from "../io/readers.ts";

const reader = new BufReader(new StringReader(test.in));
const conn = createConnMock();
const [req, err] = await readRequest(conn, reader);
```

Reply:
Add a todo referencing this: https://github.com/denoland/deno/issues/2383